### PR TITLE
Flax CLM script

### PR DESCRIPTION
**Pull request by patil-suraj**
This PR adds causal language model training script for flax.

### CODE REVIEW

File: examples/flax/language-modeling/run_clm_flax.py [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+        # summarize metrics
+        metrics = {"loss": loss}
+        if distributed:
+            metrics = jax.lax.pmean(metrics, axis_name="batch")
+        return metrics
+
+    # Create parallel version of the train and eval step
+    if distributed:
+        train_step = jax.pmap(_train_step, "batch", donate_argnums=(0,))
+        eval_step = jax.pmap(_eval_step, "batch", donate_argnums=(0,))
```
Comment (n2cholas) [inline examples/flax/language-modeling/run_clm_flax.py:546]: Can remove `donate_argnums=(0,)` for `jax.pmap(_eval_step)` since it won't be used anyway.

File: examples/flax/language-modeling/run_clm_flax.py [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+        # Create sampling rng
+        rng, input_rng = jax.random.split(rng)
+
+        # Generate an epoch by shuffling sampling indices from the train dataset
+        train_loader = data_loader(input_rng, train_dataset, train_batch_size, shuffle=True, distributed=distributed)
+        steps_per_epoch = len(train_dataset) // train_batch_size
+        # train
+        for _ in tqdm(range(steps_per_epoch), desc="Training...", position=1, leave=False):
+            batch = next(train_loader)
+            state, train_metric, dropout_rngs = train_step(state, batch, dropout_rngs)
```
Comment (n2cholas) [inline examples/flax/language-modeling/run_clm_flax.py:577]: This is good, but in my view, it's cleaner to include the `dropout_rngs` in the `TrainState`, so you don't have the input and output rng in the train loop here (since an rng really is state). Just a preference, up to you. 

Here's how I set it up in a similar script: https://colab.research.google.com/drive/1P8I5js9Yv_jGuB8BzFzcUm0suML3zFcV?usp=sharing

File: examples/flax/language-modeling/run_clm_flax.py [hunk 3]
```diff
@@ -0,0 +1,632 @@
...
+            train_metrics.append(train_metric)
+
+        train_time += time.time() - train_start
+
+        if distributed:
+            train_metric = unreplicate(train_metric)
```
Comment (n2cholas) [inline examples/flax/language-modeling/run_clm_flax.py:583]: I believe you can use `get_metrics` here like you do for `eval`.

File: examples/flax/language-modeling/run_clm_flax.py [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+
+        # Save metrics
+        if has_tensorboard and jax.process_index() == 0:
+            cur_step = epoch * (len(train_dataset) // train_batch_size)
+            write_metric(summary_writer, train_metrics, eval_metrics, train_time, cur_step, distributed)
+
+    # save last checkpoint
+    if jax.process_index() == 0:
+        if distributed:
+            params = jax.device_get(jax.tree_map(lambda x: x[0], state.params))
```
Comment (n2cholas) [inline examples/flax/language-modeling/run_clm_flax.py:625]: You can use `unreplicate` here instead of the `tree_map`.

File: examples/flax/language-modeling/run_clm_flax.py [hunk 5]
```diff
@@ -0,0 +1,632 @@
...
+        batch_idx = jax.random.permutation(rng, len(dataset))
+    else:
+        batch_idx = jnp.arange(len(dataset))
+
+    batch_idx = batch_idx[: steps_per_epoch * batch_size]  # Skip incomplete batch.
+    batch_idx = batch_idx.reshape((steps_per_epoch, batch_size))
+
+    for idx in batch_idx:
+        batch = dataset[idx]
+        batch = {k: jnp.array(v) for k, v in batch.items()}
```
Comment (n2cholas) [inline examples/flax/language-modeling/run_clm_flax.py:206]: It would be more generic to write this as `batch = jax.tree_map(jnp.asarray, dataset[idx])`.

File: examples/flax/language-modeling/run_clm_flax.py [hunk 6]
```diff
@@ -0,0 +1,632 @@
...
+
+        if distributed:
+            batch = shard(batch)
```
Comment (n2cholas) [inline examples/flax/language-modeling/run_clm_flax.py:209]: I'm curious if there would be any performance benefit if you use [`prefetch_to_device`](https://github.com/google/flax/blob/65061e6128f6695eed441acf2bfffc3b1badd318/flax/jax_utils.py#L131) here instead.

**Review by n2cholas — COMMENTED** [review]
Thanks for putting this together, it looks great overall!

File: examples/flax/language-modeling/run_clm_flax.py (hunk 2 above)
Comment (patil-suraj) [reply examples/flax/language-modeling/run_clm_flax.py]: This is much cleaner indeed!

File: examples/flax/language-modeling/run_clm_flax.py (hunk 3 above)
Comment (patil-suraj) [reply examples/flax/language-modeling/run_clm_flax.py:588]: this is a single `dict` for last train step in the epoch, so we can't use `get_metrics` here.

File: examples/flax/language-modeling/run_clm_flax.py (hunk 5 above)
Comment (patil-suraj) [reply examples/flax/language-modeling/run_clm_flax.py:213]: This requires `jax.tree_map(jnp.asarray, batch, is_leaf=lambda x: type(x) == list)`, so I would prefer `dict` comprehension

File: examples/flax/language-modeling/run_clm_flax.py (hunk 6 above)
Comment (patil-suraj) [reply examples/flax/language-modeling/run_clm_flax.py:216]: I tried this on TPU and as the docstring states, doesn't seem to make much difference on TPU. Will do a few runs on multi-gpu setup later and update the script in a follow-up PR if it gives any boost.

File: examples/flax/language-modeling/run_clm_flax.py [hunk 7]
```diff
@@ -0,0 +1,632 @@
...
+        return traverse_util.unflatten_dict(flat_mask)
+
+    # create adam optimizer
+    adamw = optax.adamw(
+        learning_rate=linear_decay_lr_schedule_fn,
+        b1=training_args.adam_beta1,
+        b2=training_args.adam_beta2,
+        eps=training_args.adam_epsilon,
+        weight_decay=training_args.weight_decay,
+    )
```
[the remark this answers is not in the document]
Comment (patil-suraj) [reply examples/flax/language-modeling/run_clm_flax.py:501]: Aha, yes!
@patrickvonplaten  In the `run_mlm_flax.py` script we are not using the `decay_mask_fn`, is that intentional?

File: examples/flax/language-modeling/run_clm_flax.py [hunk 8]
```diff
@@ -0,0 +1,637 @@
...
+        eval_dataset = lm_datasets["validation"]
+        if data_args.max_eval_samples is not None:
+            eval_dataset = eval_dataset.select(range(data_args.max_eval_samples))
+
+    # Enable tensorboard only on the master node
+    if has_tensorboard and jax.process_index() == 0:
+        summary_writer = SummaryWriter(log_dir=Path(training_args.output_dir).joinpath("logs").as_posix())
+
+    # Initialize our training
+    distributed = jax.local_device_count() > 1
```
Comment (patil-suraj) [inline examples/flax/language-modeling/run_clm_flax.py:465]: @patrickvonplaten , @n2cholas is this a better way to detect distributed training, or should we rather add a flag?

File: examples/flax/language-modeling/run_clm_flax.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply examples/flax/language-modeling/run_clm_flax.py:501]: We should probably add the mask fn to the `mlm` script as well

File: examples/flax/language-modeling/run_clm_flax.py (hunk 8 above)
Comment (patrickvonplaten) [reply examples/flax/language-modeling/run_clm_flax.py:465]: Think this is fine!

File: examples/flax/language-modeling/run_clm_flax.py [hunk 9]
```diff
@@ -0,0 +1,637 @@
...
+        transformers.utils.logging.set_verbosity_info()
+    else:
+        datasets.utils.logging.set_verbosity_error()
+        transformers.utils.logging.set_verbosity_error()
+
+    # Set the verbosity to info of the Transformers logger (on main process only):
+    logger.info(f"Training/evaluation parameters {training_args}")
+
+    # Set seed before initializing model.
+    set_seed(training_args.seed)
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:292]: Don't think that this is needed since no numpy random function is used here

File: examples/flax/language-modeling/run_clm_flax.py [hunk 10]
```diff
@@ -0,0 +1,637 @@
...
+        tokenize_function,
+        batched=True,
+        num_proc=data_args.preprocessing_num_workers,
+        remove_columns=column_names,
+        load_from_cache_file=not data_args.overwrite_cache,
+    )
+
+    if data_args.block_size is None:
+        block_size = tokenizer.model_max_length
+        if block_size > 1024:
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:402]: Hmm, not a huge fan of hard-coding 1024 here - could we maybe use `config.max_position_embeddings` instead?

File: examples/flax/language-modeling/run_clm_flax.py [hunk 11]
```diff
@@ -0,0 +1,637 @@
...
+    state = TrainState.create(apply_fn=model.__call__, params=model.params, tx=adamw, dropout_rng=dropout_rng)
+
+    def loss_fn(logits, labels):
+        shift_logits = logits[..., :-1, :]
+        shift_labels = labels[..., 1:]
+        loss = optax.softmax_cross_entropy(shift_logits, onehot(shift_labels, shift_logits.shape[-1]))
+        return loss.mean()
+
+    # Define gradient update step fn
+    def _train_step(state, batch):
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:513]: (nit) why call it `_train_step` instead of `train_step`? in the MLM script, it's called `train_step` -> maybe do the same here for consistency?

**Review by patrickvonplaten — APPROVED** [review]
LGTM - left some nits

File: examples/flax/language-modeling/run_clm_flax.py [hunk 12]
```diff
@@ -0,0 +1,633 @@
...
+        eval_step = jax.jit(_eval_step)
+
+    logger.info("***** Running training *****")
+    logger.info(f"  Num examples = {len(train_dataset)}")
+    logger.info(f"  Num Epochs = {num_epochs}")
+    logger.info(f"  Instantaneous batch size per device = {training_args.per_device_train_batch_size}")
+    logger.info(f"  Total train batch size (w. parallel & distributed) = {train_batch_size}")
+    logger.info(f"  Total optimization steps = {total_train_steps}")
+
+    train_metrics = []
```
Comment (jheek) [inline examples/flax/language-modeling/run_clm_flax.py:562]: this should go into the epoch loop. Otherwise training metrics accumulate (on device) forever and are submitted as metrics multiple times

Comment (jheek) [reply examples/flax/language-modeling/run_clm_flax.py:562]: Also you might need to periodically fetch the metrics to memory if the number of steps per epoch get very large.

File: examples/flax/language-modeling/run_clm_flax.py [hunk 13]
```diff
@@ -0,0 +1,633 @@
...
+        batch_idx = jnp.arange(len(dataset))
+
+    batch_idx = batch_idx[: steps_per_epoch * batch_size]  # Skip incomplete batch.
+    batch_idx = batch_idx.reshape((steps_per_epoch, batch_size))
+
+    for idx in batch_idx:
+        batch = dataset[idx]
+        batch = {k: jnp.array(v) for k, v in batch.items()}
+
+        if distributed:
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:214]: Actually can we get rid of the whole `if distributed` logic?

All functions `pmap`, `replicate`, ... also work on single device -> so let's just stick to that no?

File: examples/flax/language-modeling/run_clm_flax.py [hunk 14]
```diff
@@ -0,0 +1,633 @@
...
+
+    # Define eval fn
+    def _eval_step(params, batch):
+        labels = batch.pop("labels")
+        logits = model(**batch, params=params, train=False)[0]
+        loss = loss_fn(logits, labels)
+
+        # summarize metrics
+        metrics = {"loss": loss}
+        if distributed:
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:540]: for single device we should also take the mean no?

File: examples/flax/language-modeling/run_clm_flax.py [hunk 15]
```diff
@@ -0,0 +1,633 @@
...
+    )
+
+    if data_args.block_size is None:
+        block_size = tokenizer.model_max_length
+        if block_size > config.max_position_embeddings:
+            logger.warning(
+                f"The tokenizer picked seems to have a very large `model_max_length` ({tokenizer.model_max_length}). "
+                "Picking 1024 instead. You can change that default value by passing --block_size xxx."
+            )
+        block_size = 1024
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:403]:
```suggestion
            block_size = 1024
```

Comment (patrickvonplaten) [reply examples/flax/language-modeling/run_clm_flax.py:403]: Otherwise block_size is always 1024

File: examples/flax/language-modeling/run_clm_flax.py [hunk 16]
```diff
@@ -0,0 +1,614 @@
...
+        return new_state, metrics
+
+    # Define eval fn
+    def eval_step(params, batch):
+        labels = batch.pop("labels")
+        logits = model(**batch, params=params, train=False)[0]
+        loss = loss_fn(logits, labels)
+
+        # summarize metrics
+        metrics = {"loss": loss}
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:532]: Also, think this should be `loss.mean()` here

Comment (patrickvonplaten) [reply examples/flax/language-modeling/run_clm_flax.py:532]: If you compare to:         metrics = {"loss": loss.sum(), "accuracy": accuracy.sum(), "normalizer": label_mask.sum()}
